test/infrastructure/docker/exp/internal/controllers/dockermachinepool_controller.go

```diff
@@ -164,21 +179,45 @@ func (r *DockerMachinePoolReconciler) SetupWithManager(ctx context.Context, mgr
 			builder.WithPredicates(
 				predicates.ClusterUnpausedAndInfrastructureReady(ctrl.LoggerFrom(ctx)),
 			),
-		).Complete(r)
+		).Build(r)
 	if err != nil {
 		return errors.Wrap(err, "failed setting up with a controller manager")
 	}
+
+	r.controller = c
+	r.recorder = mgr.GetEventRecorderFor("dockermachinepool-controller")
+	r.externalTracker = external.ObjectTracker{
+		Controller: c,
+		Cache:      mgr.GetCache(),
+	}
+
 	return nil
 }
 
 func (r *DockerMachinePoolReconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster, machinePool *expv1.MachinePool, dockerMachinePool *infraexpv1.DockerMachinePool) error {
-	pool, err := docker.NewNodePool(ctx, r.Client, cluster, machinePool, dockerMachinePool)
+	_ = ctrl.LoggerFrom(ctx)
+
+	labelFilters := map[string]string{dockerMachinePoolLabel: dockerMachinePool.Name}
+	machines, err := docker.ListMachinesByCluster(ctx, cluster, labelFilters)
 	if err != nil {
-		return errors.Wrap(err, "failed to build new node pool")
+		return errors.Wrapf(err, "failed to list all machines in the cluster")
 	}
 
-	if err := pool.Delete(ctx); err != nil {
-		return errors.Wrap(err, "failed to delete all machines in the node pool")
+	for _, machine := range machines {
+		key := client.ObjectKey{
+			Namespace: dockerMachinePool.Namespace,
+			Name:      machine.Name(),
+		}
+		dockerMachine := &infrav1.DockerMachine{}
+		if err := r.Client.Get(ctx, key, dockerMachine); err != nil {
+			if err := machine.Delete(ctx); err != nil {
+				return errors.Wrapf(err, "failed to delete machine %s", machine.Name())
+			}
+		}
+
+		if err := r.deleteMachinePoolMachine(ctx, *dockerMachine, *machinePool); err != nil {
+			return errors.Wrapf(err, "failed to delete machine %s", dockerMachine.Name)
+		}
```

Comment:
The primary case we expect is to delete the Machine as you described to reuse the node drain/delete, but I recall we discussed what to do when the Machine doesn't exist. What we concluded was:

- If the MP controller is running, we can requeue and wait for the Machine to be created eventually because otherwise we could have a race condition.
- Otherwise, i.e. if the MP is deleted, just delete the DockerMachine directly.

I'm open to suggestions on how to clean this up a bit. I wonder if we need to delete the DockerMachines if the MP is being deleted as they are owned by the DockerMachinePool and will be cleaned up due to ownerRefs.

Reply:
First of all I think that this code in its current form lacks of comments. 
Please consider whether to add at least a comment for each method, and explain why it exists (in this case ~the comment above), which is more important to track than documenting step-by-step what a method does.

If you are looking at ideas for how to refactor this, https://github.com/fabriziopandini/cluster-api/blob/8965f4c50dd26d2336e4b4483b44738e17bfb59f/internal/controllers/topology/cluster/reconcile_state.go#L476 can probably be a viable approach.
It first computes the diff between the current and desired state, and then it starts doing all the actions to align them

Reply:
I can definitely add some comments. As for the desired state refactor, let me play around with it and see if it would be straightforward to do.